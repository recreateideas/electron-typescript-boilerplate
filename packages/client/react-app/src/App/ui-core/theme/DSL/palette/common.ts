import { Palette, ThemeColor } from '../../';

export const primary = {
    ice: {
        50: '#FFFFFF',
        100: '#EBF8FF',
        200: '#99DBFF',
        300: '#85D6FF',
        400: '#70CFFF',
        500: '#00ACFF',
        600: '#0096E0',
        700: '#005F8F',
        800: '#003652',
        900: '#001B29',
    } as ThemeColor,
    purple: {
        50: '#EFE4F9',
        100: '#D8BCF0',
        200: '#BE90E6',
        300: '#A364DB',
        400: '#9042D4',
        500: '#7C21CC',
        600: '#741DC7',
        700: '#6918C0',
        800: '#5F14B9',
        900: '#4C0BAD',
    } as ThemeColor,
    darkIndigo: {
        50: '#E4E3EA',
        100: '#BCB9CB',
        200: '#908AA8',
        300: '#645B85',
        400: '#42386B',
        500: '#211551',
        600: '#1D124A',
        700: '#180F40',
        800: '#140C37',
        900: '#0B0627',
    } as ThemeColor,
    royalBlueDark: {
        500: '#031A51',
    } as ThemeColor,
    white: '#ffffff',
};

export const secondary = {
    red: {
        50: '#FEE4EC',
        100: '#FDBBD0',
        200: '#FC8EB0',
        300: '#FB6090',
        400: '#FA3E79',
        500: '#F91C61',
        600: '#f81959',
        700: '#F7144F',
        800: '#F61145',
        900: '#F50933',
    } as ThemeColor,
    orange: {
        50: '#FFEEEA',
        100: '#FED5CB',
        200: '#FEB9A8',
        300: '#FD9D85',
        400: '#FC886B',
        500: '#FC7351',
        600: '#FC6B4A',
        700: '#FB6040',
        800: '#FB5637',
        900: '#FA4327',
    } as ThemeColor,
    green: {
        50: '#E2F9F4',
        100: '#B7EFE4',
        200: '#87E4D2',
        300: '#57D9C0',
        400: '#33D1B3',
        500: '#0FC9A5',
        600: '#0DC39D',
        700: '#0BBC93',
        800: '#08B58A',
        900: '#04A979',
    } as ThemeColor,
    neon: {
        50: '#E3FAFE',
        100: '#B8F3FD',
        200: '#89ECFB',
        300: '#59E4F9',
        400: '#36DEF8',
        500: '#12D8F7',
        600: '#10D4F6',
        700: '#0DCEF5',
        800: '#0AC8F3',
        900: '#05BFF1',
    } as ThemeColor,
    yellow: {
        50: '#FFFAE9',
        100: '#FFF4C7',
        200: '#FFECA2',
        300: '#FEE47C',
        400: '#FEDF60',
        500: '#FED944',
        600: '#FED53E',
        700: '#FECF35',
        800: '#FECA2D',
        900: '#FDC01F',
    } as ThemeColor,
};

export const tertiary = {
    blue: {
        50: '#E8ECFD',
        100: '#C7D0FA',
        200: '#A1B0F7',
        300: '#7B90F3',
        400: '#5F79F1',
        500: '#4361EE',
        600: '#3D59EC',
        700: '#344FE9',
        800: '#2C45E7',
        900: '#1E33E2',
    } as ThemeColor,
    pink: {
        50: '#FDE6F9',
        100: '#F9C2EF',
        200: '#F699E5',
        300: '#F270DB',
        400: '#EF51D3',
        500: '#EC32CB',
        600: '#EA2DC6',
        700: '#E726BE',
        800: '#E41FB8',
        900: '#DF13AC',
    } as ThemeColor,
    brown: {
        50: '#F2EAE7',
        100: '#DFCBC3',
        200: '#CAA99B',
        300: '#B48673',
        400: '#A46C55',
        500: '#945237',
        600: '#8C4B31',
        700: '#81412A',
        800: '#773823',
        900: '#652816',
    } as ThemeColor,
    grey: {
        50: '#EFEFEF',
        100: '#D7D7D7',
        200: '#BDBDBD',
        300: '#A3A3A3',
        400: '#8F8F8F',
        500: '#7B7B7B',
        600: '#737373',
        700: '#686868',
        800: '#5E5E5E',
        900: '#4B4B4B',
    } as ThemeColor,
};

export const brand = {
    primary: primary.ice,
} as Palette['brand'];

export const fonts = {
    info: brand.primary,
    success: secondary.green[500],
    warn: secondary.orange[500],
    error: secondary.red[500],
};
