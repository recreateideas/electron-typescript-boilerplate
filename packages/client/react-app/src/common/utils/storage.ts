/* Retrieve layout from local storage. */
export const getFromLS = (key: string): string | null =>
    window.localStorage?.getItem(key) ? JSON.parse(window.localStorage.getItem(key) || '') : null;

/* Save layout to local storage. */
export const saveToLS = (key: string, value: string | number | object) =>
    window.localStorage && window.localStorage.setItem(key, JSON.stringify(value));
