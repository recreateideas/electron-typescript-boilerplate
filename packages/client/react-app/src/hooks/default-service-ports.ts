import { IServicePorts } from '../typings';

export const defaultServicePorts = {
    'local-service': 2999,
} as IServicePorts;
