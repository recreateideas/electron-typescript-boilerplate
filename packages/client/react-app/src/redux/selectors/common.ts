import { Selector } from 'react-redux';
import { IReduxStore } from '../store';
import { IServicePorts } from '../../typings';

const servicePorts: Selector<IReduxStore, IServicePorts | undefined> = (state) =>
    state.appState.common.servicePorts;

const isElectron: Selector<IReduxStore, boolean | undefined> = (state) =>
    state.appState.common.isElectron;

export { isElectron, servicePorts };
