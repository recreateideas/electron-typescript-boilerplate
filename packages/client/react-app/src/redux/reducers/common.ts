import { IReduxStore, IReduxAction } from '../store';
import initialState from '../store/initialState';
import { types } from '../actions/common';

type SliceCommon = IReduxStore['common'];

const common = (state: SliceCommon = initialState.common, action: IReduxAction): SliceCommon => {
    const { type, data } = action;
    switch (type) {
        case types.SET_IS_ELECTRON:
            return {
                ...state,
                isElectron: data.isElectron,
            };
        case types.GET_SERVICE_PORTS_SUCCESS:
            return {
                ...state,
                servicePorts: data.ports,
            };
        default:
            return state;
    }
};

export default common;
