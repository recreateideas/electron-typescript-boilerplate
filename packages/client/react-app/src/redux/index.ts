import { useDispatch } from 'react-redux';
import * as selectors from './selectors';
import * as actions from './actions';
import { useSelector } from './custom-hooks';

export * from './store';

export { useDispatch, actions, useSelector, selectors };
