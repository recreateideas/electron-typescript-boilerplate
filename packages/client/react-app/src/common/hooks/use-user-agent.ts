import { useState, useEffect, useMemo } from 'react';
import { UserAgent } from 'src/redux';
//@ts-ignore
import useMobileDetect from 'use-mobile-detect-hook';

export const useUserAgent = (): UserAgent | undefined => {
    const { isMobile, isDesktop, isAndroid, isIos } = useMobileDetect();
    const [ua, setUA] = useState<UserAgent | undefined>();
    useEffect(() => {
        const getUA = () => {
            const breakpoint = 620;
            const userAgent: UserAgent = {
                isMobile: isMobile() || window.innerWidth <= breakpoint,
                isDesktop: isDesktop(),
                isAndroid: isAndroid(),
                isIos: isIos(),
            };
            setUA(userAgent);
        };
        window.addEventListener('resize', getUA);
        getUA();
        return () => {
            window.removeEventListener('resize', getUA);
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    return useMemo(() => ua, [ua?.isMobile]);
};
