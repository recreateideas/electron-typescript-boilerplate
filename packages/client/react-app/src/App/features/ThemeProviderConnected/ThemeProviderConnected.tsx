import { useEffect, useLayoutEffect } from 'react';
import PropTypes from 'prop-types';
import { isEqual } from 'lodash';
import { useDispatch, selectors, useSelector, actions } from '../../../redux';
import { getFromLS } from '../../../common/utils';
import { ThemeProvider, ThemeMode } from '../../ui-core';
import { useUserAgent } from '../../../common/hooks';

interface IProps {
    children: React.ReactNode;
}

export const ThemeProviderConnected = (props: IProps) => {
    const dispatch = useDispatch();
    const { children } = props;
    const { theme: themeSelectors } = selectors;
    const {
        theme: { setUserAgent, setThemeMode },
    } = actions;
    const modeLS: string | null = getFromLS('mode');
    const mode = useSelector(themeSelectors.mode);
    console.log(mode);
    const userAgentStore = useSelector(themeSelectors.userAgent);
    const userAgent = useUserAgent();
    const browserMode = (isSystemDarkMode: boolean) => (isSystemDarkMode ? 'dark' : 'light');

    useEffect(() => {
        if (userAgent && !isEqual(userAgentStore, userAgent)) {
            dispatch(setUserAgent(userAgent));
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [userAgentStore, userAgent]);

    useLayoutEffect(() => {
        if (modeLS) {
            dispatch(setThemeMode(modeLS as ThemeMode));
            return;
        }
        const { matches: isSystemDarkMode } = window.matchMedia('(prefers-color-scheme: dark)');
        const mode = browserMode(isSystemDarkMode);
        dispatch(setThemeMode(mode));
        window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', (event) => {
            console.log(event.toString());
            const mode = browserMode(event.matches);
            dispatch(setThemeMode(mode));
        });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return mode ? (
        <ThemeProvider {...{ mode }}>{children}</ThemeProvider>
    ) : (
        <div className="no-theme">{children}</div>
    );
};

ThemeProviderConnected.propTypes = {
    children: PropTypes.any,
};
