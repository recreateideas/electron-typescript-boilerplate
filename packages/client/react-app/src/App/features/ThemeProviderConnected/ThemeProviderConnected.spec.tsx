/* eslint-disable import/first */
import { ThemeProviderConnected } from './ThemeProviderConnected';
jest.mock('../../../redux');
import * as redux from '../../../redux';
import { TestProvider } from '../../../tests';
import { initialState } from '../../../redux';
jest.mock('../../../common/hooks');
import * as hooks from '../../../common/hooks';
jest.mock('../../../common/utils');
import * as utils from '../../../common/utils';

const { mountWithProvider } = TestProvider(initialState);

describe('<ThemeProviderConnected/>', () => {
    const userAgent = {
        isMobile: true,
        isDesktop: false,
        isAndroid: false,
        isIos: true,
    };
    it('should render', () => {
        const dispatchSpy = jest.fn();
        jest.spyOn(redux, 'useDispatch').mockReturnValue(dispatchSpy);
        const root = mountWithProvider(
            <ThemeProviderConnected>some-children</ThemeProviderConnected>
        );
        expect(root).toMatchSnapshot();
    });
    it('should render the children wrapped by no-theme div if no theme mode is found', () => {
        const dispatchSpy = jest.fn();
        jest.spyOn(redux, 'useDispatch').mockReturnValue(dispatchSpy);
        const root = mountWithProvider(
            <ThemeProviderConnected>
                <div className="children">some-children</div>
            </ThemeProviderConnected>
        );
        const children = root.find('.no-theme .children');
        expect(children.length).toBe(1);
    });

    it('should render the children wrapped in ThemeProvider if a theme mode is found', () => {
        const dispatchSpy = jest.fn();
        jest.spyOn(redux, 'useDispatch').mockReturnValue(dispatchSpy);
        jest.spyOn(redux, 'useDispatch').mockReturnValue(dispatchSpy);
        jest.spyOn(redux, 'useSelector')
            .mockReturnValueOnce('light')
            .mockReturnValueOnce(userAgent);
        const root = mountWithProvider(
            <ThemeProviderConnected>
                <div className="children">some-children</div>
            </ThemeProviderConnected>
        );
        const children = root.find('ThemeProvider .children');
        expect(children.length).toBe(1);
    });
    it('should dispatch setUserAgent with the userAgent found', () => {
        const dispatchSpy = jest.fn();
        jest.spyOn(redux, 'useDispatch').mockReturnValue(dispatchSpy);
        jest.spyOn(redux, 'useSelector')
            .mockReturnValueOnce('light')
            .mockReturnValueOnce(undefined);
        jest.spyOn(hooks, 'useUserAgent')
            .mockReturnValueOnce(userAgent)
            .mockReturnValueOnce(userAgent);
        mountWithProvider(
            <ThemeProviderConnected>
                <div />
            </ThemeProviderConnected>
        );
        expect(redux.actions.theme.setUserAgent).toHaveBeenCalledWith(userAgent);
    });

    it('should dispatch setThemeMode with the mode found in localStorage', () => {
        const dispatchSpy = jest.fn();
        jest.spyOn(redux, 'useDispatch').mockReturnValue(dispatchSpy);
        jest.spyOn(redux, 'useSelector').mockReturnValueOnce('light');
        jest.spyOn(utils, 'getFromLS').mockReturnValue('dark');

        mountWithProvider(
            <ThemeProviderConnected>
                <div />
            </ThemeProviderConnected>
        );
        expect(redux.actions.theme.setThemeMode).toHaveBeenCalledWith('dark');
    });

    it('should dispatch setThemeMode 2 times, once after having found the system dark mode', () => {
        Object.defineProperty(window, 'matchMedia', {
            writable: true,
            value: (query: string) => ({
                matches: false,
                addEventListener: jest.fn().mockImplementation((_, fn) => fn({ matches: false })),
            }),
        });
        const dispatchSpy = jest.fn();
        jest.spyOn(redux, 'useDispatch').mockReturnValue(dispatchSpy);
        jest.spyOn(redux, 'useSelector').mockReturnValueOnce('light');
        jest.spyOn(utils, 'getFromLS').mockReturnValue(null);

        mountWithProvider(
            <ThemeProviderConnected>
                <div />
            </ThemeProviderConnected>
        );
        expect(redux.actions.theme.setThemeMode).toHaveBeenCalledTimes(2);
        expect(redux.actions.theme.setThemeMode).toHaveBeenNthCalledWith(2, 'light');
    });

    it('should dispatch setThemeMode 2 times, once after having found the system dark mode - 2', () => {
        Object.defineProperty(window, 'matchMedia', {
            writable: true,
            value: (query: string) => ({
                matches: true,
                addEventListener: jest.fn().mockImplementation((_, fn) => fn({ matches: true })),
            }),
        });
        const dispatchSpy = jest.fn();
        jest.spyOn(redux, 'useDispatch').mockReturnValue(dispatchSpy);
        jest.spyOn(redux, 'useSelector').mockReturnValueOnce('light');
        jest.spyOn(utils, 'getFromLS').mockReturnValue(null);

        mountWithProvider(
            <ThemeProviderConnected>
                <div />
            </ThemeProviderConnected>
        );
        expect(redux.actions.theme.setThemeMode).toHaveBeenCalledTimes(2);
        expect(redux.actions.theme.setThemeMode).toHaveBeenNthCalledWith(2, 'dark');
    });
});
