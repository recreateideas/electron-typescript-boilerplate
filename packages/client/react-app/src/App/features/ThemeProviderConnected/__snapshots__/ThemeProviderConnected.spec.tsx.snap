// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThemeProviderConnected/> should render 1`] = `
<div
  className="no-theme"
>
  some-children
</div>
`;
